src/selectors_vm/stack.rs

```diff
@@ -1,27 +1,32 @@
-use super::SelectorState;
-use super::program::AddressRange;
 use super::ast::NthChild;
-use crate::html::{LocalName, Namespace, Tag};
+use super::program::AddressRange;
+use super::SelectorState;
+use crate::html::{EsiComment, EsiInclude, LocalName, Namespace, Tag};
 use crate::memory::{LimitedVec, MemoryLimitExceededError, SharedMemoryLimiter};
 // use hashbrown for raw entry, switch back to std once it stablizes there
-use hashbrown::{HashSet, HashMap, hash_map::RawEntryMut};
+use encoding_rs::Encoding;
+use hashbrown::{hash_map::RawEntryMut, HashMap, HashSet};
 use std::fmt::Debug;
-use std::hash::{Hash, Hasher, BuildHasher};
+use std::hash::{BuildHasher, Hash, Hasher};
 
 #[inline]
-fn is_void_element(local_name: &LocalName) -> bool {
+fn is_void_element(local_name: &LocalName, encoding: &'static Encoding) -> bool {
     // NOTE: fast path for the most commonly used elements
     if tag_is_one_of!(*local_name, [Div, A, Span, Li]) {
         return false;
     }
 
-    tag_is_one_of!(
+    if tag_is_one_of!(
         *local_name,
         [
             Area, Base, Basefont, Bgsound, Br, Col, Embed, Hr, Img, Input, Keygen, Link, Meta,
             Param, Source, Track, Wbr
         ]
-    )
+    ) {
+        return true;
+    }
+
+    byte_tag_is_one_of!(local_name, encoding, [EsiInclude, EsiComment])
```

Comment:
You don't need to pass around encoding and decode the tag name: it's all ASCII and all the encodings are ASCII-safe. Just compare local name's bytes, e.g.:
```
if let LocalName::Bytes(bytes) = local_name {
    if bytes == b"esi:include" || bytes == b"esi:comment" {
         return true;
    }
}

```